# Task 2: BLEU score
The official BLEU evaluation script published by NIST is
`mteval-v11a-cmufix_b.pl` Run the script with the argument `-h` and
look at the "help" output:

In [3]:
export LANGUAGE=C.UTF-8
export LC_ALL=C.UTF-8

In [4]:
mteval-v11a-cmufix_b.pl -h

MT evaluation scorer began on 2022 Jun 2 at 09:20:44
command line:  /usr/local/bin/mteval-v11a-cmufix_b.pl -h


Usage: /usr/local/bin/mteval-v11a-cmufix_b.pl [-h] -r <ref_file> -s src_file -t <tst_file>

Description:  This Perl script evaluates MT system performance.

Required arguments:
  -r <ref_file> is a file containing the reference translations for
      the documents to be evaluated.
  -s <src_file> is a file containing the source documents for which
      translations are to be evaluated
  -t <tst_file> is a file containing the translations to be evaluated

Optional arguments:
  -c preserves upper-case alphabetic characters
  -b generate BLEU scores only
  -n generate NIST scores only
  -d detailed output flag used in conjunction with "-b" or "-n" flags:
         0 (default) for system-level score only
         1 to include document-level scores
         2 to include segment-level scores
         3 to include ngram-level scores
  -m normalization method:
	      n (default) NIST

: 255

## (a)
For each evaluation run (e. g. for each test file) use the
following arguments:  
* i. Calculate BLEU and NIST score (`-b`, `-n`)

In [5]:
mteval-v11a-cmufix_b.pl -r /opt/data/wmt10-xlats/ref/wmt10-newssyscombtest2010-ref.en.sgm -s /opt/data/wmt10-xlats/ref/wmt10-newssyscombtest2010-src.de.sgm -t /opt/data/wmt10-xlats/newssyscombtest2010.de-en.kit.sgm -n

MT evaluation scorer began on 2022 Jun 2 at 09:20:56
command line:  /usr/local/bin/mteval-v11a-cmufix_b.pl -r /opt/data/wmt10-xlats/ref/wmt10-newssyscombtest2010-ref.en.sgm -s /opt/data/wmt10-xlats/ref/wmt10-newssyscombtest2010-src.de.sgm -t /opt/data/wmt10-xlats/newssyscombtest2010.de-en.kit.sgm -n
NormalizationText method: None
Use closest ref to calculate LP for BLEU
  Evaluation of any-to-en translation using:
    src set "test2009" (94 docs, 2034 segs)
    ref set "test2009" (1 refs)
    tst set "test2009" (1 systems)

NIST LenPen=1.0000
NIST score = 5.9367  for system kit

# ------------------------------------------------------------------------

Individual N-gram scoring
        1-gram   2-gram   3-gram   4-gram   5-gram   6-gram   7-gram   8-gram   9-gram
        ------   ------   ------   ------   ------   ------   ------   ------   ------
 NIST:  4.6886   1.0651   0.1679   0.0131   0.0020   0.0006   0.0001   0.0000   0.0000  "kit"


# ----------------------------------------

In [6]:
mteval-v11a-cmufix_b.pl -r /opt/data/wmt10-xlats/ref/wmt10-newssyscombtest2010-ref.en.sgm -s /opt/data/wmt10-xlats/ref/wmt10-newssyscombtest2010-src.de.sgm -t /opt/data/wmt10-xlats/newssyscombtest2010.de-en.kit.sgm -b

MT evaluation scorer began on 2022 Jun 2 at 09:20:59
command line:  /usr/local/bin/mteval-v11a-cmufix_b.pl -r /opt/data/wmt10-xlats/ref/wmt10-newssyscombtest2010-ref.en.sgm -s /opt/data/wmt10-xlats/ref/wmt10-newssyscombtest2010-src.de.sgm -t /opt/data/wmt10-xlats/newssyscombtest2010.de-en.kit.sgm -b
NormalizationText method: None
Use closest ref to calculate LP for BLEU
  Evaluation of any-to-en translation using:
    src set "test2009" (94 docs, 2034 segs)
    ref set "test2009" (1 refs)
    tst set "test2009" (1 systems)

Bleu LenPen=1.0000

BLEU score = 0.1822 for system kit

# ------------------------------------------------------------------------

Individual N-gram scoring
        1-gram   2-gram   3-gram   4-gram   5-gram   6-gram   7-gram   8-gram   9-gram
        ------   ------   ------   ------   ------   ------   ------   ------   ------
 BLEU:  0.4937   0.2412   0.1285   0.0721   0.0412   0.0242   0.0145   0.0087   0.0053  "kit"

# -----------------------------------------

* ii.  Use the text normalization method "BLEU" (`-m b` option)

In [7]:
mteval-v11a-cmufix_b.pl -r /opt/data/wmt10-xlats/ref/wmt10-newssyscombtest2010-ref.en.sgm -s /opt/data/wmt10-xlats/ref/wmt10-newssyscombtest2010-src.de.sgm -t /opt/data/wmt10-xlats/newssyscombtest2010.de-en.kit.sgm -m b

MT evaluation scorer began on 2022 Jun 2 at 09:21:02
command line:  /usr/local/bin/mteval-v11a-cmufix_b.pl -r /opt/data/wmt10-xlats/ref/wmt10-newssyscombtest2010-ref.en.sgm -s /opt/data/wmt10-xlats/ref/wmt10-newssyscombtest2010-src.de.sgm -t /opt/data/wmt10-xlats/newssyscombtest2010.de-en.kit.sgm -m b
NormalizationText method: Bleu
Use closest ref to calculate LP for BLEU
  Evaluation of any-to-en translation using:
    src set "test2009" (94 docs, 2034 segs)
    ref set "test2009" (1 refs)
    tst set "test2009" (1 systems)

NIST LenPen=1.0000
Bleu LenPen=1.0000
NIST score = 6.8816 LP: 1.0000 BLEU score = 0.2397 Prec: 0.2397 LP: 1.0000 for system kit

# ------------------------------------------------------------------------

Individual N-gram scoring
        1-gram   2-gram   3-gram   4-gram   5-gram   6-gram   7-gram   8-gram   9-gram
        ------   ------   ------   ------   ------   ------   ------   ------   ------
 NIST:  5.2220   1.3805   0.2476   0.0265   0.0051   0.0009   0

* iii.  Calculate case-sensitive scores (`-c`)

In [8]:
mteval-v11a-cmufix_b.pl -r /opt/data/wmt10-xlats/ref/wmt10-newssyscombtest2010-ref.en.sgm -s /opt/data/wmt10-xlats/ref/wmt10-newssyscombtest2010-src.de.sgm -t /opt/data/wmt10-xlats/newssyscombtest2010.de-en.kit.sgm -m b -c

MT evaluation scorer began on 2022 Jun 2 at 09:21:08
command line:  /usr/local/bin/mteval-v11a-cmufix_b.pl -r /opt/data/wmt10-xlats/ref/wmt10-newssyscombtest2010-ref.en.sgm -s /opt/data/wmt10-xlats/ref/wmt10-newssyscombtest2010-src.de.sgm -t /opt/data/wmt10-xlats/newssyscombtest2010.de-en.kit.sgm -m b -c
NormalizationText method: Bleu
Use closest ref to calculate LP for BLEU
  Evaluation of any-to-en translation using:
    src set "test2009" (94 docs, 2034 segs)
    ref set "test2009" (1 refs)
    tst set "test2009" (1 systems)

NIST LenPen=1.0000
Bleu LenPen=1.0000
NIST score = 6.6975 LP: 1.0000 BLEU score = 0.2288 Prec: 0.2288 LP: 1.0000 for system kit

# ------------------------------------------------------------------------

Individual N-gram scoring
        1-gram   2-gram   3-gram   4-gram   5-gram   6-gram   7-gram   8-gram   9-gram
        ------   ------   ------   ------   ------   ------   ------   ------   ------
 NIST:  5.1633   1.2854   0.2218   0.0228   0.0042   0.0008 

-   What is the BLEU and NIST score for the four systems?

-   What is the Precision and Length Penalty of the BLEU scores?

-   Does the automatic evaluation look like what you expected?

-   How is the BLEU score calculated from the n-gram matches?
  - Geometric means of the 1-gram to $n$-gram scores times the length penalty
  
-   What is the difference between the individual and the
    cumulative score?
  - The cumulative $n$-gram score for is effectively the $n$-gram BLEU score, e.g. the cumulative 4-gram score is the same as the standard BLEU score  
    (it takes into account the $1..n$-gram scores)
  - Individual $n$-gram scores take into account the $n$-grams for this one $n$ only

|System |NIST   |BLEU   |Prec.  |LP    |
|-------|-------|-------|-------|------|
|DFKI   |4.8896 |0.1243 |0.1243 |1.0000|
|JHU    |5.5791 |0.1542 |0.1659 |0.9790|
|KIT    |5.9367 |0.1822 |0.1822 |1.0000|
|RWTH   |5.9293 |0.1829 |0.1829 |1.0000|


## (b)
Evaluate our system (KIT) with the case-sensitive option.

-   Do you notice any difference in the score? Explain the
    difference if any.


In [9]:
mteval-v11a-cmufix_b.pl -r /opt/data/wmt10-xlats/ref/wmt10-newssyscombtest2010-ref.en.sgm -s /opt/data/wmt10-xlats/ref/wmt10-newssyscombtest2010-src.de.sgm -t /opt/data/wmt10-xlats/newssyscombtest2010.de-en.kit.sgm -b -m b -c | grep ^BLEU
mteval-v11a-cmufix_b.pl -r /opt/data/wmt10-xlats/ref/wmt10-newssyscombtest2010-ref.en.sgm -s /opt/data/wmt10-xlats/ref/wmt10-newssyscombtest2010-src.de.sgm -t /opt/data/wmt10-xlats/newssyscombtest2010.de-en.kit.sgm -b -m b | grep ^BLEU

BLEU score = 0.2288 for system kit
BLEU score = 0.2397 for system kit


## (c)
Evaluate our system (KIT) only for document-level BLEU score:

```bash
mteval-v11a-cmufix_b.pl \
    -r /opt/data/wmt10-xlats/ref/wmt10-newssyscombtest2010-ref.en.sgm \
    -s /opt/data/wmt10-xlats/ref/wmt10-newssyscombtest2010-src.de.sgm \
    -t ./newssyscombtest2010.de-en.kit.sgm \
    -b -m b -d 1
```


In [10]:
mteval-v11a-cmufix_b.pl \
    -r /opt/data/wmt10-xlats/ref/wmt10-newssyscombtest2010-ref.en.sgm \
    -s /opt/data/wmt10-xlats/ref/wmt10-newssyscombtest2010-src.de.sgm \
    -t /opt/data/wmt10-xlats/newssyscombtest2010.de-en.kit.sgm \
    -b -m b -d 1

MT evaluation scorer began on 2022 Jun 2 at 09:21:22
command line:  /usr/local/bin/mteval-v11a-cmufix_b.pl -r /opt/data/wmt10-xlats/ref/wmt10-newssyscombtest2010-ref.en.sgm -s /opt/data/wmt10-xlats/ref/wmt10-newssyscombtest2010-src.de.sgm -t /opt/data/wmt10-xlats/newssyscombtest2010.de-en.kit.sgm -b -m b -d 1
NormalizationText method: Bleu
Use closest ref to calculate LP for BLEU
  Evaluation of any-to-en translation using:
    src set "test2009" (94 docs, 2034 segs)
    ref set "test2009" (1 refs)
    tst set "test2009" (1 systems)

Bleu LenPen=1.0000
BLEU score using   4-grams = 0.1480 for system "kit" on document "abces/2009/12/10/200783" (13 segments, 473 words)
Bleu LenPen=1.0000
BLEU score using   4-grams = 0.1853 for system "kit" on document "abces/2009/12/10/200784" (16 segments, 611 words)
Bleu LenPen=1.0000
BLEU score using   4-grams = 0.2336 for system "kit" on document "abces/2009/12/11/200957" (9 segments, 259 words)
Bleu LenPen=1.0000
BLEU score using   4-grams = 0.1404 f

Bleu LenPen=1.0000
BLEU score using   4-grams = 0.2795 for system "kit" on document "lesechos/2009/12/11/218659" (13 segments, 362 words)
Bleu LenPen=0.9397
BLEU score using   4-grams = 0.1778 for system "kit" on document "lesechos/2009/12/12/218769" (13 segments, 354 words)
Bleu LenPen=0.9906
BLEU score using   4-grams = 0.1601 for system "kit" on document "lesechos/2009/12/12/218771" (15 segments, 529 words)
Bleu LenPen=0.9729
BLEU score using   4-grams = 0.1319 for system "kit" on document "lidovky.cz/2009/12/10/75424" (23 segments, 510 words)
Bleu LenPen=1.0000
BLEU score using   4-grams = 0.1502 for system "kit" on document "lidovky.cz/2009/12/10/75429" (47 segments, 754 words)
Bleu LenPen=0.9001
BLEU score using   4-grams = 0.0903 for system "kit" on document "lidovky.cz/2009/12/10/75456" (12 segments, 285 words)
Bleu LenPen=1.0000
BLEU score using   4-grams = 0.2328 for system "kit" on document "lidovky.cz/2009/12/10/75462" (24 segments, 622 words)
Bleu LenPen=1.0000
BLEU score 

-   Why does one of the documents have a zero score?
  - Document `elmundo/2009/12/11/10121` only has 71 words in total
  - Due to short document length there might not be a single matching 4-gram
  - Product of $n$-gram scores becomes 0

In [11]:
cat /opt/data/wmt10-xlats/ref/wmt10-newssyscombtest2010-ref.en.sgm |
    grep -F -A8 'elmundo/2009/12/11/10121' |
    perl -ne '/<seg id="\d+"> (.*)<\/seg>/ && print $1, "\n"'

echo

cat /opt/data/wmt10-xlats/newssyscombtest2010.de-en.kit.sgm |
    grep -F -A8 'elmundo/2009/12/11/10121' |
    perl -ne '/<seg id="\d+"> (.*)<\/seg>/ && print $1, "\n"'

'Bag Girl'
Close to Karl Lagerfeld, the image of the Coco Cocoon line and occasional 'starlet' of the catwalk,the singer lavishes her bad-girl image with as much contempt as she has Chanel bags.
More than a 'Bad Girl' or 'It Girl', she is a'Bag Girl'.

Bag girl '
The singer, a friend of Finland by Karl storage area, is the face of the cosmetics line Coco Cocoon and a Gelegenheitsstarlett during the Modenschauen ' maison ', they presented their rude face with the same such stubbornness as their pockets of Chanel.
Much more than "bad girl" or "it girl", she is a "Bag girl."


Find the document with the highest and the lowest BLEU score
(ignore the document with zero score) and read the translation.

In [12]:
mteval-v11a-cmufix_b.pl \
    -r /opt/data/wmt10-xlats/ref/wmt10-newssyscombtest2010-ref.en.sgm \
    -s /opt/data/wmt10-xlats/ref/wmt10-newssyscombtest2010-src.de.sgm \
    -t /opt/data/wmt10-xlats/newssyscombtest2010.de-en.kit.sgm \
    -b -m b -d 1 |
    grep "^BLEU score using" |
    sort -k6 -n

BLEU score using   4-grams = 0.0000 for system "kit" on document "elmundo/2009/12/11/10121" (3 segments, 71 words)
BLEU score using   4-grams = 0.0903 for system "kit" on document "lidovky.cz/2009/12/10/75456" (12 segments, 285 words)
BLEU score using   4-grams = 0.1058 for system "kit" on document "cincodias/2009/12/10/53255" (5 segments, 175 words)
BLEU score using   4-grams = 0.1136 for system "kit" on document "elmundo/2009/12/11/10127" (12 segments, 479 words)
BLEU score using   4-grams = 0.1213 for system "kit" on document "elmundo/2009/12/10/10101" (9 segments, 265 words)
BLEU score using   4-grams = 0.1217 for system "kit" on document "lidovky.cz/2009/12/10/75470" (22 segments, 540 words)
BLEU score using   4-grams = 0.1273 for system "kit" on document "elmundo/2009/12/11/10124" (19 segments, 596 words)
BLEU score using   4-grams = 0.1295 for system "kit" on document "lidovky.cz/2009/12/10/75482" (25 segments, 514 words)
BLEU score using   4-grams = 0.1311 for system "kit" on d

BLEU score using   4-grams = 0.2680 for system "kit" on document "spiegel/2009/12/12/73257" (127 segments, 2543 words)
BLEU score using   4-grams = 0.2759 for system "kit" on document "lidovky.cz/2009/12/10/75483" (15 segments, 339 words)
BLEU score using   4-grams = 0.2795 for system "kit" on document "lesechos/2009/12/11/218659" (13 segments, 362 words)
BLEU score using   4-grams = 0.2824 for system "kit" on document "washpost/2009/12/12/79432" (28 segments, 743 words)
BLEU score using   4-grams = 0.2836 for system "kit" on document "frankfurter-rundschau/2009/12/10/9949" (33 segments, 789 words)
BLEU score using   4-grams = 0.2837 for system "kit" on document "lesechos/2009/12/11/218644" (20 segments, 677 words)
BLEU score using   4-grams = 0.2844 for system "kit" on document "lidovky.cz/2009/12/10/75486" (30 segments, 781 words)
BLEU score using   4-grams = 0.2886 for system "kit" on document "washpost/2009/12/12/79407" (51 segments, 1264 words)
BLEU score using   4-grams = 0.2891 

-   Can we explain the difference in scores?
  - Performance is very much dependent on the document type/domain
-   Try to provide a speculation about the reasons for the
    different quality of the translations.
  - Documents with bad performance tend to be from non-German and non-English news
  - Could result in greater number of out-of-vocabulary words

# Task 3: TER Score
TER stands for Translation Error Rate. The TER is the ratio of
errors in the hypothesis to the number of words in the reference
translation. The mistakes are composed of insertions, deletions,
substitutions and shifts of words.

## (a) 
Calculate the translation error rate for the first sentence of
the translation of our system (KIT):

-   How many editing actions do you need to restore the
    reference from the hypothesis?

| ref:  | lack of snow |   in   | mountains | causes problems |  for hoteliers   |
|-------|--------------|--------|-----------|-----------------|------------------|
| dist: | 0            |   2    |    0      |       2         |        3         |
| hyp:  | lack of snow | on the | mountains |  has bedeviled  | the hotel owners |

Execute the script to calculate the TER score for this
translation

```bash
tercom_v6b.pl -h <hypothesis-file> -r <reference-file> -i sgm -N
```

In [13]:
# the hypothesis directory needs to be writable so copy it to /tmp
cp /opt/data/wmt10-xlats/newssyscombtest2010.de-en.kit.sgm /tmp
tercom_v6b.pl -h /tmp/newssyscombtest2010.de-en.kit.sgm -r /opt/data/wmt10-xlats/ref/wmt10-newssyscombtest2010-ref.en.sgm -i sgm -N > /dev/null

## (b)
While the program is running, you can take a look at the file
**newssyscombtest2010.de-en.kit.sgm.sys.pra** (it is being
generated by the script).

-   Did you calculate the TER of the first sentence correctly?

In [14]:
head -n 10 /tmp/newssyscombtest2010.de-en.kit.sgm.sys.pra

Sentence ID: idnes.cz/2009/12/11/76492-1
Best Ref: lack of snow in mountains causes problems for hoteliers
Orig Hyp: lack of snow on the mountains has bedeviled the hotel owners

REF:  lack of snow ** IN  mountains *** CAUSES    PROBLEMS FOR   HOTELIERS
HYP:  lack of snow ON THE mountains HAS BEDEVILED THE      HOTEL OWNERS   
EVAL:              I  S             I   S         S        S     S        
SHFT:                                                                     
TER Score:  77.78 (  7.0/  9.0)



-   Is it possible for the error rate to exceed 100%? How can
    this happen?
  - Yes, it's possible.  
    We divide by the reference length so a long hypothesis might result in a large number of deletions,
    resulting in an error count higher than the reference length
  - See part (c) below for an example

In [15]:
head -n 120 /tmp/newssyscombtest2010.de-en.kit.sgm.sys.pra

Sentence ID: idnes.cz/2009/12/11/76492-1
Best Ref: lack of snow in mountains causes problems for hoteliers
Orig Hyp: lack of snow on the mountains has bedeviled the hotel owners

REF:  lack of snow ** IN  mountains *** CAUSES    PROBLEMS FOR   HOTELIERS
HYP:  lack of snow ON THE mountains HAS BEDEVILED THE      HOTEL OWNERS   
EVAL:              I  S             I   S         S        S     S        
SHFT:                                                                     
TER Score:  77.78 (  7.0/  9.0)


Sentence ID: idnes.cz/2009/12/11/76492-2
Best Ref: it's not only krkonoše ski-lift operators who are worried by these bare slopes .
Orig Hyp: make scneelose pisten sorgenn not only the giant mountains-local residents .

REF:  IT'S   not only   KRKONOšE SKI-LIFT OPERATORS WHO       ARE    WORRIED   BY  THESE BARE            SLOPES    .
HYP:  **** [ not only ] ********* ******** MAKE      SCNEELOSE PISTEN SORGENN @ THE GIANT MOUNTAINS-LOCAL RESIDENTS .
EVAL: D                 D       

 HYP:  AND IF   they call US TO @ THEM IMMEDIATELY ,     WHETHER WE   [ ask ] BELIEVE THAT THIS YEAR'S CHRISTMAS ON  snow or "   MATSCH " WILL BE LED  martin jandura BY THE OPERATORS OF   the spindl . info information PORTAL ,    .

Shift [,] 13 words right
 REF:  "   WHEN they call ** ,  THE  FIRST         THING   THEY ask IS      IF   WE   THINK  THEY'LL   SEE snow or * MUD      ,   " **** ** SAYS martin jandura ** ,   WHO       RUNS the spindl . info information WEB    SITE .
 HYP:  AND IF   they call US TO THEM IMMEDIATELY @ WHETHER WE   ask BELIEVE THAT THIS YEAR'S CHRISTMAS ON  snow or " MATSCH [ , ] " WILL BE LED  martin jandura BY THE OPERATORS OF   the spindl . info information PORTAL ,    .

Shift [we] 4 words right
 REF:  "   WHEN they call ,  THE FIRST THING       THEY      ask ******* IS   IF     we   THINK  THEY'LL   SEE snow or * MUD    , " **** ** SAYS martin jandura ** ,   WHO       RUNS the spindl . info information WEB    SITE .
 HYP:  AND IF   they call US TO  THEM 

## (c)
Find the sentence ID "idnes.cz/2009/12/11/76492-24" in the file
**newssyscombtest2010.de-en.kit.sgm.sys.pra**.

The source sentence was:
`Wir liegen etwa um fünf Prozent besser als im Vorjahr.`

-   What is the TER of the sentence?

In [16]:
grep -F -A9 "idnes.cz/2009/12/11/76492-24" /tmp/newssyscombtest2010.de-en.kit.sgm.sys.pra

Sentence ID: idnes.cz/2009/12/11/76492-24
Best Ref: we're up about 5 percent .
Orig Hyp: we are about five per cent better than in the previous year .

REF:  WE'RE UP  about **** *** **** ****** **** ** *** 5        PERCENT .
HYP:  WE    ARE about FIVE PER CENT BETTER THAN IN THE PREVIOUS YEAR    .
EVAL: S     S         I    I   I    I      I    I  I   S        S        
SHFT:                                                                    
TER Score: 183.33 ( 11.0/  6.0)



-   What error rate would you have given (as a human being) if
    you do not look at the reference but at the source sentence
    for comparison?
    
-   How can you partly tackle this problem?
  - Better Normalization of the sentences (e.g. `we're` -> `we &apos; re`) will result in less errors in this case
  - More on that in the next exercise

## (d)
In the file newssyscombtest2010.de-en.kit.sgm.sys.sum\_doc you
will find the TER scores for all documents and for the entire
test set.

-   What is the TER score for the KIT system?

In [17]:
#perl -ne 's/Sent Id/Send_Id/; print if ($. == 4 || $. == '"$(wc -l < /tmp/newssyscombtest2010.de-en.kit.sgm.sys.sum_doc)"')' < /tmp/newssyscombtest2010.de-en.kit.sgm.sys.sum_doc | column -t
sed -ne '/Sent Id/ {s/Sent Id/Sent_Id/; p}; $p' < /tmp/newssyscombtest2010.de-en.kit.sgm.sys.sum_doc | column -t

Sent_Id  |  Ins   |  Del   |  Sub    |  Shft  |  WdSh  |  NumEr    |  NumWd      |  TER
TOTAL    |  5261  |  3878  |  17010  |  4050  |  5285  |  30199.0  |  51023.000  |  59.187


-   You can repeat the TER computation with the other 3 systems.
  - No thanks

# Task 4: TER Score - Character-based
Look at the following files. The files contain the same sentences
but in a different format.

*References:*  
`/opt/data/wmt16-ende/newstest2015-ende-ref.de.sgm`  
`/opt/data/wmt16-ende/newstest2015-ende-ref.de`  

*Hypotheses:*  
`/opt/data/wmt16-ende/newstest2015-ende.kit.de.sgm`  
`/opt/data/wmt16-ende/newstest2015-ende.kit.de`  

Copy them into your directory.
```bash
cp /opt/data/wmt16-ende/newstest2015-ende-ref.de.sgm .
cp /opt/data/wmt16-ende/newstest2015-ende-ref.de .
cp /opt/data/wmt16-ende/newstest2015-ende.kit.de.sgm .
cp /opt/data/wmt16-ende/newstest2015-ende.kit.de .
```

## (a)
Calculate the translation quality with character-based TER using
the following command:

```bash
CharacTER.py \
    -r newstest2015-ende-ref.de \
    -o newstest2015-ende.kit.de \
    -v > wmt15_kit_charTER.Log
```

In [18]:
CharacTER.py \
    -r /opt/data/wmt16-ende/newstest2015-ende-ref.de \
    -o /opt/data/wmt16-ende/newstest2015-ende.kit.de \
    -v |
    sed -n '1,16p; 17s/^.*$/.../; /^CharacTER/! p'

CharacTER of sentence 1 is 0.6889
CharacTER of sentence 2 is 0.6516
CharacTER of sentence 3 is 0.4266
CharacTER of sentence 4 is 0.8615
CharacTER of sentence 5 is 0.7176
CharacTER of sentence 6 is 0.5632
CharacTER of sentence 7 is 0.3012
CharacTER of sentence 8 is 0.6558
CharacTER of sentence 9 is 0.5960
CharacTER of sentence 10 is 0.4958
CharacTER of sentence 11 is 0.9327
CharacTER of sentence 12 is 0.4464
CharacTER of sentence 13 is 0.1642
CharacTER of sentence 14 is 0.3913
CharacTER of sentence 15 is 0.6026
CharacTER of sentence 16 is 0.4270
...
0.5674459255624126
Shift cost:
0.08672359368481158


Now look at the output file:

-   What is the char-TER score of each sentence? Which sentence
    is the best and which one is the worst?

In [19]:
CharacTER.py \
    -r /opt/data/wmt16-ende/newstest2015-ende-ref.de \
    -o /opt/data/wmt16-ende/newstest2015-ende.kit.de \
    -v |
    grep "^CharacTER" |
    sort -k6 -n |
    sed -n '1p; $p'

CharacTER of sentence 1029 is 0.0000
CharacTER of sentence 316 is 1.8000


In [20]:
echo "Best:"
sed -n -s '1029p' /opt/data/wmt16-ende/newstest2015-ende{-ref,.kit}.de
echo
echo "Worst:"
sed -n -s '316p' /opt/data/wmt16-ende/newstest2015-ende{-ref,.kit}.de

Best:
Asbest ist schließlich ein Produkt der Vergangenheit.
Asbest ist schließlich ein Produkt der Vergangenheit.

Worst:
Beweg dich!
Move!


## (b)
Calculate the translation quality with word-based TER using the
following command:

```bash
tercom_v6b.pl \
    -r newstest2015-ende-ref.de.sgm \
    -h newstest2015-ende.kit.de.sgm -i sgm -N
```

Look at the following file
`newstest2015-ende.kit.de.sgm.sys.pra` and
`newstest2015-ende.kit.de.sgm.sys.sum_doc`.

-   What is the word-TER score of each document? Which sentence
    is the best and which one is the worst?


In [21]:
cp /opt/data/wmt16-ende/newstest2015-ende.kit.de.sgm /tmp
tercom_v6b.pl \
    -r /opt/data/wmt16-ende/newstest2015-ende-ref.de.sgm \
    -h /tmp/newstest2015-ende.kit.de.sgm -i sgm -N > /dev/null

In [22]:
cat /tmp/newstest2015-ende.kit.de.sgm.sys.sum_doc | sed '1,3d; s/Sent Id/Sent_Id/' | grep -v '^-*$' | column -t

Sent_Id                         |  Ins   |  Del   |  Sub    |  Shft  |  WdSh  |  NumEr    |  NumWd      |  TER
1012-bbc                        |  10    |  7     |  34     |  7     |  15    |  58.0     |  95.000     |  61.053
1035-news.com.au                |  7     |  23    |  55     |  6     |  7     |  91.0     |  179.000    |  50.838
107-stv.tv                      |  1     |  12    |  44     |  8     |  20    |  65.0     |  133.000    |  48.872
1141-thenation.com              |  33    |  80    |  280    |  50    |  58    |  443.0    |  816.000    |  54.289
1149-reuters                    |  23    |  64    |  164    |  33    |  39    |  284.0    |  484.000    |  58.678
1158-news.com.au                |  16    |  59    |  152    |  28    |  34    |  255.0    |  434.000    |  58.756
1209-telegraph                  |  16    |  19    |  115    |  25    |  33    |  175.0    |  275.000    |  63.636
12325-dw                        |  25    |  23    |  73     |  11    |  16    |  132.0    |

7896-oe24.at                    |  133   |  57    |  317    |  83    |  122   |  590.0    |  932.000    |  63.305
7917-mainpost.de                |  61    |  21    |  202    |  39    |  58    |  323.0    |  465.000    |  69.462
8108-jungewelt.de               |  103   |  30    |  222    |  44    |  57    |  399.0    |  601.000    |  66.389
8152-mittelbayerische.de        |  14    |  9     |  47     |  5     |  6     |  75.0     |  177.000    |  42.373
837-telegraph                   |  15    |  61    |  121    |  28    |  35    |  225.0    |  466.000    |  48.283
899-reuters                     |  27    |  57    |  138    |  34    |  46    |  256.0    |  502.000    |  50.996
918-telegraph                   |  25    |  36    |  114    |  17    |  20    |  192.0    |  263.000    |  73.004
940-latimes                     |  49    |  112   |  350    |  66    |  78    |  577.0    |  1044.000   |  55.268
949-news.com.au                 |  14    |  43    |  107    |  16    |  22    |  180.0  